src/services/item/plugins/published/plugins/search/meilisearch.ts

```diff
@@ -0,0 +1,397 @@
+
+import {
+  DocumentItemExtra,
+  ItemTagType,
+  ItemType,
+  LocalFileItemExtra,
+  MimeTypes,
+  S3FileItemExtra,
+} from '@graasp/sdk';
+
+import { FILE_STORAGE_ROOT_PATH } from '../../../../../../utils/config';
+import { Repositories, buildRepositories } from '../../../../../../utils/repositories';
+import FileService from '../../../../../file/service';
+import { Item } from '../../../../entities/Item';
+import { readPdfContent } from '../../../../utils';
+import { stripHtml } from '../../../validation/utils';
+import { ItemPublishedRepository } from '../../repositories/itemPublished';
+
+const ACTIVE_INDEX = 'itemIndex';
+const ROTATING_INDEX = 'itemIndex_tmp'; // Used when reindexing
+
+type ALLOWED_INDICES = typeof ACTIVE_INDEX | typeof ROTATING_INDEX;
+
+// Make index configuration typesafe
+const SEARCHABLE_ATTRIBUTES: (keyof IndexItem)[] = ['name', 'description', 'content'];
+const DISPLAY_ATTRIBUTES: (keyof IndexItem)[] = [
+  'id',
+  'name',
+  'description',
+  'type',
+  'content',
+  'createdAt',
+  'updatedAt',
+  'categories',
+  'isPublishedRoot',
+  'isHidden',
+];
+const FILTERABLE_ATTRIBUTES: (keyof IndexItem)[] = ['categories', 'isPublishedRoot', 'isHidden'];
+const TYPO_TOLERANCE: TypoTolerance = {
+  enabled: true,
+  minWordSizeForTypos: {
+    oneTypo: 4,
+    twoTypos: 6,
+  },
+};
+
+/*
+ * Handle search index business logic with Meilisearch
+ * Ideally we try to keep the public method idempotent. You can "delete" unexisting items and indexing work for first indexation and for updates.
+ */
+export class MeiliSearchWrapper {
+  private meilisearchClient: MeiliSearch;
+  indexDictionary: Record<string, Index<IndexItem>> = {};
+  fileService: FileService;
+  db: DataSource;
+  logger: FastifyBaseLogger;
+  constructor(
+    db: DataSource,
+    meilisearchConnection: MeiliSearch,
+    fileService: FileService,
+    logger: FastifyBaseLogger,
+  ) {
+    this.meilisearchClient = meilisearchConnection;
+    this.fileService = fileService;
+    this.db = db;
+    this.logger = logger;
+
+    // create index in the background if it doesn't exist
+    this.getIndex();
+  }
+
+  private removeHTMLTags(s: string): string {
+    if (s === null) return '';
+    return stripHtml(s);
+  }
+
+  private async logTaskCompletion(task: EnqueuedTask, itemName: string) {
+    const finishedTask = await (await this.getIndex()).waitForTask(task.taskUid);
+    if (task.status === TaskStatus.TASK_SUCCEEDED) {
+      this.logger.info(
+        `Meilisearch task completed: item name => ${itemName}, task type => ${task.type}, result: Successful task`,
+      );
+    } else {
+      this.logger.info(
+        `Meilisearch task completed: item name => ${itemName}, task type => ${task.type}, result: ${finishedTask.error}`,
+      );
+    }
+  }
+
+  // Lazily create or get the index at first request and cache it in the dictionary
+  private async getIndex(name: ALLOWED_INDICES = ACTIVE_INDEX): Promise<Index<IndexItem>> {
```

Comment:
Add comment as you explained it to me :D

Reply:
👌